docs/GFXReconstruct Architectural Principles.md

```diff
@@ -0,0 +1,220 @@
+
+Capture should avoid changing the content of the application’s graphics commands.  Some exceptions are necessary in order to save information required for replay or modify data so that replay is possible.
+
+* Implications:  
+  * It *may* be acceptable for the user to *request* an annotated or altered capture.  (e.g. A user may wish to save the GPU buffers used by drawing commands at capture time.  However, the preferred technique for altering the application’s graphics command stream and saving additional information for a specialization of GFXR is to make an additional upchain or downchain layer separate from GFXR. For APIs without a formal third-party layering procedure (e.g. DirectX 12), we may implement a plugin architecture for pre- or post- function and method call alteration of commands.)  
+  * Combined with the above principle “A capture file should contain as much as possible of all the graphics API calls made by the application”, a corollary is that API calls modified in capture should be stored in their original form in the capture file; a MetaData Command may encode additional information to signal to the replay application to also modify the calls on replay.  
+* Benefits:  
+  * GFXR limits the differences between running the application with and without the capture layer enabled are limited.  GFXR increases confidence that replay of the capture file will closely match the application’s execution without the capture layer.  
+* Exceptions:  
+  * Capture enables the Vulkan device features bufferDeviceAddressCaptureReplay, rayTracingPipelineShaderGroupHandleCaptureReplay, and accelerationStructureCaptureReplay during capture of Vulkan applications (and in replay) to allow replay on the same device and driver of ray-tracing workloads.  
+    * Capture may alter function calls (e.g. add USAGE flags to memory creation) that enable the use of those device features  
+  * Before a trim range and, in particular, immediately before the first function within a trim range:  
+    * GFXReconstruct may make API calls to query state, read memory contents, and synchronize with the underlying implementation.  
+    * GFXReconstruct may remove some functions that pertained only to previous frame contents that are not included in the trim range.  An example is if vkQueuePresentKHR was omitted from state setup, then vkWaitForFences  
+  * Additionally, any changes in the future that may alter original application behavior should be configurable with a capture-time option. 
+
+## REPLAY 
+
+**Replay without user intervention in the**same** **environment** as capture is an explicit goal. **
+
+Replay on the same GPU model as capture, on the same operating system, with the same driver revision, under the same or lower load conditions without additional user-provided options**is** an explicit goal.
+
```

Comment:
No mention of portable replay as a goal at all?

Reply:
This document is meant to make strong statements against which we can measure a change to the code.

While LunarG and contributors to the project do put a **lot** of effort into attempting to replay on implementations that differ from the capture-time implementation, I think I would stop short of saying it is an explicit **architectural principle**.

I think portable replay is more like "we want it if it's possible practically within the engineering resource available".  Maybe it's worth saying something like "while portable replay is not a strong design principle, it is highly desired, therefore capture must not make portable replay impossible" or some such.  I'll think about how to rephrase that.